sw/device/silicon_creator/lib/cert/cert.c

```diff
@@ -0,0 +1,227 @@
+static keymgr_binding_value_t sealing_binding_value = {.data = {0}};
+static attestation_public_key_t curr_pubkey = {.x = {0}, .y = {0}};
+static uint8_t uds_tbs_buffer[kUdsMaxTbsSizeBytes];
+static uds_sig_values_t uds_cert_params = {
+    .tbs = uds_tbs_buffer,
+    .tbs_size = kUdsMaxTbsSizeBytes,
+};
+static uint8_t cdi_0_tbs_buffer[kCdi0MaxTbsSizeBytes];
+static cdi_0_sig_values_t cdi_0_cert_params = {
+    .tbs = cdi_0_tbs_buffer,
+    .tbs_size = kCdi0MaxTbsSizeBytes,
+};
+static uint8_t cdi_1_tbs_buffer[kCdi1MaxTbsSizeBytes];
+static cdi_1_sig_values_t cdi_1_cert_params = {
+    .tbs = cdi_1_tbs_buffer,
+    .tbs_size = kCdi1MaxTbsSizeBytes,
+};
+
+static_assert(kAttestMeasurementSizeInBytes == 32,
+              "The attestation measurement size should equal the size of the "
+              "keymgr binding registers.");
+
+/**
+ * Returns true if debug (JTAG) access is exposed in the current LC state.
+ */
+static bool is_debug_exposed(void) {
+  lifecycle_state_t lc_state = lifecycle_state_get();
+  if (lc_state == kLcStateProd || lc_state == kLcStateProdEnd) {
+    return false;
+  }
+  return true;
+}
+
+status_t gen_uds_keys_and_cert(manuf_cert_perso_data_in_t *perso_data_in,
+                               hmac_digest_t *uds_pubkey_id, uint8_t *cert,
+                               size_t *cert_size) {
+  // Generate the UDS key.
+  TRY(keymgr_state_check(kKeymgrStateInit));
+  keymgr_advance_state();
+  TRY(keymgr_state_check(kKeymgrStateCreatorRootKey));
+  TRY(otbn_boot_attestation_keygen(kUdsAttestationKeySeed,
+                                   kUdsKeymgrDiversifier, &curr_pubkey));
+  TRY(otbn_boot_attestation_key_save(kUdsAttestationKeySeed,
+                                     kUdsKeymgrDiversifier));
+
+  // Generate the key ID.
+  hmac_sha256(&curr_pubkey, kAttestationPublicKeyCoordBytes * 2, uds_pubkey_id);
+
+  // Generate the (unendorsed) certificate.
+  uds_tbs_values_t uds_cert_tbs_params = {
+      // TODO(#19455): include OTP measurements in attestation keygen / cert.
+      .otp_creator_sw_cfg_hash = NULL,
+      .otp_creator_sw_cfg_hash_size = 0,
+      .otp_owner_sw_cfg_hash = NULL,
+      .otp_owner_sw_cfg_hash_size = 0,
+      .otp_hw_cfg0_hash = NULL,
+      .otp_hw_cfg0_hash_size = 0,
+      .debug_flag = is_debug_exposed(),
+      .creator_pub_key_id = (unsigned char *)uds_pubkey_id->digest,
+      .creator_pub_key_id_size = kCertKeyIdSizeInBytes,
+      .auth_key_key_id = perso_data_in->auth_key_key_id,
+      .auth_key_key_id_size = kCertKeyIdSizeInBytes,
+      .creator_pub_key_ec_x = (unsigned char *)curr_pubkey.x,
+      .creator_pub_key_ec_x_size = kAttestationPublicKeyCoordBytes,
+      .creator_pub_key_ec_y = (unsigned char *)curr_pubkey.y,
+      .creator_pub_key_ec_y_size = kAttestationPublicKeyCoordBytes,
+  };
+  TRY(uds_build_tbs(&uds_cert_tbs_params, uds_cert_params.tbs,
+                    &uds_cert_params.tbs_size));
+  TRY(uds_build_cert(&uds_cert_params, cert, cert_size));
```

Comment:
Here it is building the certificate without computing the actual signature (which will be all zeroes I guess?)

Reply:
yes, fixed (see above)